File: packages/sourcepoint_unified_cmp_android/android/src/main/kotlin/de/thekorn/sourcepoint/unified/cmp/SourcePointUnifiedCmpData.kt
package de.thekorn.sourcepoint.unified.cmp

import HostAPIActionType
import HostAPICampaignType
import HostAPICampaignsEnv
import HostAPIConsentAction
import HostAPIConsentStatus
import HostAPIGDPRConsent
import HostAPIGDPRPurposeGrants
import HostAPIGranularState
import HostAPIGranularStatus
import HostAPIMessageLanguage
import HostAPIMessageType
import HostAPIPMTab
import HostAPISPConsent
import com.sourcepoint.cmplibrary.data.network.model.optimized.ConsentStatus
import com.sourcepoint.cmplibrary.data.network.model.optimized.GranularState
import com.sourcepoint.cmplibrary.data.network.util.CampaignsEnv
import com.sourcepoint.cmplibrary.exception.CampaignType
import com.sourcepoint.cmplibrary.model.ConsentAction
import com.sourcepoint.cmplibrary.model.MessageLanguage
import com.sourcepoint.cmplibrary.model.PMTab
import com.sourcepoint.cmplibrary.model.exposed.ActionType
import com.sourcepoint.cmplibrary.model.exposed.GDPRPurposeGrants
import com.sourcepoint.cmplibrary.model.exposed.MessageType
import com.sourcepoint.cmplibrary.model.exposed.SPConsents
import com.sourcepoint.cmplibrary.model.exposed.SPGDPRConsent
import com.sourcepoint.cmplibrary.model.exposed.toWebViewConsentsJsonObject

fun GDPRPurposeGrants.toHostAPIPurposeGrants() = HostAPIGDPRPurposeGrants(
    granted = granted,
    purposeGrants = purposeGrants.mapKeys { it.key }.mapValues { it.value }
)

fun SPGDPRConsent.toHostAPIGDPRConsent() = HostAPIGDPRConsent(
    uuid = consent.uuid,
    tcData = consent.tcData.mapKeys { it.key }.mapValues { it.value?.toString() },
    grants = consent.grants.mapKeys { it.key }.mapValues { it.value.toHostAPIPurposeGrants() },
    apply = consent.applies,
    euconsent = consent.euconsent,
    consentStatus = consent.consentStatus?.toHostAPIConsentStatus(),
    acceptedCategories = consent.acceptedCategories
)

fun GranularState.toHostAPIGranularState() = when (this) {
    GranularState.ALL -> HostAPIGranularState.ALL
    GranularState.SOME -> HostAPIGranularState.SOME
    GranularState.NONE -> HostAPIGranularState.NONE
    GranularState.EMPTY_VL -> HostAPIGranularState.EMPTY_VL
}

fun ConsentStatus.GranularStatus.toHostAPIGranularStatus() = HostAPIGranularStatus(
    defaultConsent = defaultConsent,
    previousOptInAll = previousOptInAll,
    purposeConsent = purposeConsent?.toHostAPIGranularState(),
    purposeLegInt = purposeLegInt?.toHostAPIGranularState(),
    vendorConsent = vendorConsent?.toHostAPIGranularState(),
    vendorLegInt = vendorLegInt?.toHostAPIGranularState()
)

fun ConsentStatus.toHostAPIConsentStatus() = HostAPIConsentStatus(
    consentedAll = consentedAll,
    consentedToAny = consentedToAny,
    granularStatus = granularStatus?.toHostAPIGranularStatus(),
    hasConsentData = hasConsentData,
    rejectedAny = rejectedAny,
    rejectedLI = rejectedLI,
    legalBasisChanges = legalBasisChanges,
    vendorListAdditions = vendorListAdditions
)

fun SPConsents.toHostAPISPConsent() = HostAPISPConsent(
    gdpr = gdpr?.toHostAPIGDPRConsent(),
    webConsents = this.toWebViewConsentsJsonObject().toString()
)

fun ConsentAction.toHostAPIConsentAction() = HostAPIConsentAction(
    actionType = actionType.toHostAPIActionType(),
    campaignType = campaignType.toHostAPICampaignType(),
    pubData = pubData.toString(),
    customActionId = customActionId
)

fun CampaignType.toHostAPICampaignType() = when (this) {
    CampaignType.CCPA -> HostAPICampaignType.CCPA
    CampaignType.GDPR -> HostAPICampaignType.GDPR
    CampaignType.USNAT -> HostAPICampaignType.USNAT
}

fun ActionType.toHostAPIActionType() = when (this) {
    ActionType.ACCEPT_ALL -> HostAPIActionType.ACCEPT_ALL
    ActionType.CUSTOM -> HostAPIActionType.CUSTOM
    ActionType.GET_MSG_ERROR -> HostAPIActionType.GET_MSG_ERROR
    ActionType.GET_MSG_NOT_CALLED -> HostAPIActionType.GET_MESSAGE_NOT_CALLED
    ActionType.MSG_CANCEL -> HostAPIActionType.MSG_CANCEL
    ActionType.PM_DISMISS -> HostAPIActionType.PM_DISMISS
    ActionType.REJECT_ALL -> HostAPIActionType.REJECT_ALL
    ActionType.SAVE_AND_EXIT -> HostAPIActionType.SAVE_AND_EXIT
    ActionType.SHOW_OPTIONS -> HostAPIActionType.SHOW_OPTIONS
    ActionType.UNKNOWN -> HostAPIActionType.UNKNOWN
}

fun HostAPIMessageLanguage.toMessageLanguage() = when (this) {
    HostAPIMessageLanguage.ENGLISH -> MessageLanguage.ENGLISH
    HostAPIMessageLanguage.GERMAN -> MessageLanguage.GERMAN
    HostAPIMessageLanguage.SPANISH -> MessageLanguage.SPANISH
    HostAPIMessageLanguage.FRENCH -> MessageLanguage.FRENCH
    HostAPIMessageLanguage.ITALIAN -> MessageLanguage.ITALIAN
    HostAPIMessageLanguage.DUTCH -> MessageLanguage.DUTCH
}

fun HostAPICampaignsEnv.toCampaignsEnv() = when (this) {
    HostAPICampaignsEnv.PUBLIC -> CampaignsEnv.PUBLIC
    HostAPICampaignsEnv.STAGE -> CampaignsEnv.STAGE
}

fun HostAPIPMTab.toPMTab() = when (this) {
    HostAPIPMTab.PURPOSES -> PMTab.PURPOSES
    HostAPIPMTab.DEFAULTS -> PMTab.DEFAULT
    HostAPIPMTab.VENDORS -> PMTab.VENDORS
    HostAPIPMTab.FEATURES -> PMTab.FEATURES
}

fun HostAPICampaignType.toCampaignType() = when (this) {
    HostAPICampaignType.CCPA -> CampaignType.CCPA
    HostAPICampaignType.GDPR -> CampaignType.GDPR
    HostAPICampaignType.USNAT -> CampaignType.USNAT
}

fun HostAPIMessageType.toMessageType() = when (this) {
    HostAPIMessageType.MOBILE -> MessageType.MOBILE
    HostAPIMessageType.OTT -> MessageType.OTT
    HostAPIMessageType.LEGACY_OTT -> MessageType.LEGACY_OTT
}

File: packages/sourcepoint_unified_cmp_ios/ios/sourcepoint_unified_cmp_ios/Sources/sourcepoint_unified_cmp_ios/SourcepointUnifiedCmpData.swift
import ConsentViewController

enum NotImplementedError: Error {
  case notImplemented
}

extension SPGDPRVendorGrant {
  func toHostAPIPurposeGrants() -> HostAPIGDPRPurposeGrants {
    HostAPIGDPRPurposeGrants(granted: granted, purposeGrants: purposeGrants)
  }
}

extension ConsentStatus {
  func toHostAPIConsentStatus() -> HostAPIConsentStatus {
    HostAPIConsentStatus(
      consentedAll: consentedAll,
      consentedToAny: consentedToAny,
      granularStatus: nil, // NOTE: granularStatus is not public in the ios sdk
      hasConsentData: nil, // NOTE: hasConsentData is not public in the ios sdk
      rejectedAny: rejectedAny,
      rejectedLI: rejectedLI,
      legalBasisChanges: legalBasisChanges,
      vendorListAdditions: vendorListAdditions
    )
  }
}

extension SPConsent {
  func toHostAPIGDPRConsent() throws -> HostAPIGDPRConsent {
    if let value = self as? SPConsent<SPGDPRConsent> {
      let tcData: [String?: String?]? = value.consents!.tcfData?.dictionaryValue?
        .reduce(into: [:]) { result, pair in
          let value: String? =
            (pair.value as? String)
              ?? (pair.value as? CustomStringConvertible)?.description
          result[pair.key] = value
        }

      let grants: [String?: HostAPIGDPRPurposeGrants?]? = value.consents?.vendorGrants
        .reduce(into: [:]) { result, pair in
          let value: SPGDPRVendorGrant = pair.value as SPGDPRVendorGrant
          result[pair.key] = value.toHostAPIPurposeGrants()
        }

      return HostAPIGDPRConsent(
        uuid: value.consents!.uuid,
        tcData: tcData,
        grants: grants,
        euconsent: value.consents!.euconsent,
        acceptedCategories: value.consents!.acceptedCategories,
        apply: value.consents!.applies,
        consentStatus: value.consents!.consentStatus.toHostAPIConsentStatus()
      )
    } else {
      throw NotImplementedError.notImplemented
    }
  }
}

func encodeWebConsents(webConsents: SPWebConsents) -> String? {
  guard let consentsData = try? JSONEncoder().encode(webConsents) else { return nil }
  return String(data: consentsData, encoding: .utf8)
}

extension SPUserData {
  func toHostAPISPConsent() -> HostAPISPConsent {
    HostAPISPConsent(
      gdpr: try? gdpr?.toHostAPIGDPRConsent(),
      webConsents: encodeWebConsents(webConsents: webConsents)
    )
  }
}

extension SPGDPRConsent {
  func toHostAPIGDPRConsent() -> HostAPIGDPRConsent {
    let tcData: [String?: String?]? = tcfData?.dictionaryValue?
      .reduce(into: [:]) { result, pair in
        let value: String? = (pair.value as? String) ??
          (pair.value as? CustomStringConvertible)?.description
        result[pair.key] = value
      }

    let grants: [String?: HostAPIGDPRPurposeGrants?]? = vendorGrants
      .reduce(into: [:]) { result, pair in
        result[pair.key] = pair.value.toHostAPIPurposeGrants()
      }

    return HostAPIGDPRConsent(
      uuid: uuid,
      tcData: tcData,
      grants: grants,
      euconsent: euconsent,
      acceptedCategories: acceptedCategories,
      apply: applies,
      consentStatus: consentStatus.toHostAPIConsentStatus()
    )
  }

  func toHostAPISPConsent() -> HostAPISPConsent {
    HostAPISPConsent(
      gdpr: toHostAPIGDPRConsent(),
      webConsents: nil
    )
  }
}

extension HostAPICampaignsEnv {
  func toSPCampaignEnv() -> SPCampaignEnv {
    switch self {
    case .stage:
      SPCampaignEnv.Stage
    case .publicEnv:
      SPCampaignEnv.Public
    }
  }
}

extension HostAPIMessageLanguage {
  func toSPMessageLanguage() -> SPMessageLanguage {
    switch self {
    case .german:
      SPMessageLanguage.German
    case .english:
      SPMessageLanguage.English
    case .french:
      SPMessageLanguage.French
    case .italian:
      SPMessageLanguage.Italian
    case .spanish:
      SPMessageLanguage.Spanish
    case .dutch:
      SPMessageLanguage.Dutch
    }
  }
}

extension HostAPIPMTab {
  func toSPPrivacyManagerTab() -> SPPrivacyManagerTab {
    switch self {
    case .purposes:
      SPPrivacyManagerTab.Purposes
    case .defaults:
      SPPrivacyManagerTab.Default
    case .vendors:
      SPPrivacyManagerTab.Vendors
    case .features:
      SPPrivacyManagerTab.Features
    }
  }
}

extension SPAction {
  func toHostAPIConsentAction() -> HostAPIConsentAction {
    HostAPIConsentAction(
      actionType: (try? type.toHostAPIActionType()) ?? HostAPIActionType.unknown,
      pubData: String(describing: publisherData),
      campaignType: campaignType.toHostAPICampaignType(),
      customActionId: customActionId
    )
  }
}

extension SPCampaignType {
  func toHostAPICampaignType() -> HostAPICampaignType {
    switch self {
    case .ccpa:
      return HostAPICampaignType.ccpa
    case .gdpr:
      return HostAPICampaignType.gdpr
    case .usnat:
      return HostAPICampaignType.usnat
    case .ios14:
      return HostAPICampaignType.ios14
    case .globalcmp:
      return HostAPICampaignType.globalcmp
    case .preferences:
      return HostAPICampaignType.preferences
    default:
      return HostAPICampaignType.unknown
    }
  }
}

/// FIXME(thekorn) the anddroid and the ios sdk are actually diverging here.
///   ios is using newer and more granular action types, whereas
///   android is still using the old ones
///   once android is catching up we need to unify
extension SPActionType {
  func toHostAPIActionType() throws -> HostAPIActionType {
    switch self {
    case .SaveAndExit:
      HostAPIActionType.saveAndExit
    case .PMCancel:
      HostAPIActionType.msgCancel
    case .Custom:
      HostAPIActionType.custom
    case .AcceptAll:
      HostAPIActionType.acceptAll
    case .ShowPrivacyManager:
      HostAPIActionType.showOptions
    case .RejectAll:
      HostAPIActionType.rejectAll
    case .Dismiss:
      HostAPIActionType.pmDismiss
    case .RequestATTAccess:
      throw NotImplementedError.notImplemented
    case .IDFAAccepted:
      throw NotImplementedError.notImplemented
    case .IDFADenied:
      throw NotImplementedError.notImplemented
    case .Unknown:
      HostAPIActionType.unknown
    }
  }
}

extension SPError {
  func toHostAPISPError() -> HostAPISPError {
    HostAPISPError(spCode: spCode, description: description)
  }
}
